Paintroid/src/main/java/org/catrobat/paintroid/tools/helper/JavaFillAlgorithm.kt

```diff
@@ -0,0 +1,214 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.paintroid.tools.helper
+
+import android.graphics.Bitmap
+import android.graphics.Point
+import androidx.annotation.VisibleForTesting
+import java.util.LinkedList
+import java.util.Queue
+
+private const val UP = true
+private const val DOWN = false
+
+class JavaFillAlgorithm : FillAlgorithm {
+    @VisibleForTesting
+    lateinit var pixels: Array<IntArray>
+
+    @VisibleForTesting
+    lateinit var clickedPixel: Point
+
+    @VisibleForTesting
+    lateinit var ranges: Queue<Range>
+
+    @VisibleForTesting
+    @JvmField
+    var targetColor = 0
+
+    @VisibleForTesting
+    @JvmField
+    var replacementColor = 0
+
+    @VisibleForTesting
+    @JvmField
+    var colorToleranceThresholdSquared = 0
+
+    private lateinit var filledPixels: Array<BooleanArray>
+    private var considerTolerance = false
+    private lateinit var bitmap: Bitmap
+    private var width = 0
+    private var height = 0
+
+    override fun setParameters(
+        bitmap: Bitmap,
+        clickedPixel: Point,
+        targetColor: Int,
+        replacementColor: Int,
+        colorToleranceThreshold: Float
+    ) {
+        this.bitmap = bitmap
+        width = bitmap.width
+        height = bitmap.height
+        pixels = Array(bitmap.height) { IntArray(bitmap.width) }
+        for (i in 0 until height) {
+            this.bitmap.getPixels(pixels[i], 0, width, 0, i, width, 1)
+        }
+        filledPixels = Array(bitmap.height) { BooleanArray(bitmap.width) }
+        this.clickedPixel = clickedPixel
+        this.targetColor = targetColor
+        this.replacementColor = replacementColor
+        ranges = LinkedList()
+        colorToleranceThresholdSquared = (colorToleranceThreshold * colorToleranceThreshold).toInt()
+        considerTolerance = colorToleranceThreshold > 0
+    }
+
+    override fun performFilling() {
+        var range = generateRangeAndReplaceColor(
+            clickedPixel.y, clickedPixel.x, UP
+        )
+        ranges.add(range)
+        ranges.add(Range(range.line, range.start, range.end, DOWN))
+        var row: Int
+        while (!ranges.isEmpty()) {
+            ranges.poll()?.let {
+                range = it
+            }
+            if (range.direction == UP) {
+                row = range.line - 1
+                if (row >= 0) {
+                    checkRangeAndGenerateNewRanges(range, row, UP)
+                }
+            } else {
+                row = range.line + 1
+                if (row < height) {
+                    checkRangeAndGenerateNewRanges(range, row, DOWN)
+                }
+            }
+        }
+    }
+
+    private fun checkCell(row: Int, col: Int): Boolean =
```

Comment:
Best Practice: One does not have to look into the function to know what it does. 😉 
```suggestion
    private fun shouldCellBeFilled(row: Int, col: Int): Boolean =
```
Or something similar. Whatever looks more appealing to you 😄

Reply:
done :)